File: iris.scala
import scala.io.Source
import scala.collection.mutable.ListBuffer
import scala.math.sqrt
import breeze.linalg.{DenseMatrix, eig}

// Load the Iris dataset into a ListBuffer
val irisData = ListBuffer.empty[Array[Double]]
for (line <- Source.fromFile("/path/to/iris.csv").getLines().drop(1)) {
  val fields = line.split(",")
  val values = fields.take(4).map(_.toDouble)
  irisData += values
}

// Calculate the mean, median, and standard deviation of each variable for each species
def calculateMeansBySpecies(data: List[Array[Double]]): Map[String, Array[Double]] = {
  val speciesData = separateDataBySpecies(data)
  speciesData.mapValues { data =>
    val means = for (i <- data.head.indices) yield {
      val col = data.map(_(i))
      col.sum / col.length
    }
    means.toArray
  }
}

def calculateMediansBySpecies(data: List[Array[Double]]): Map[String, Array[Double]] = {
  val speciesData = separateDataBySpecies(data)
  speciesData.mapValues { data =>
    val medians = for (i <- data.head.indices) yield {
      val col = data.map(_(i)).sorted
      if (col.length % 2 == 0) {
        (col(col.length / 2 - 1) + col(col.length / 2)) / 2
      } else {
        col(col.length / 2)
      }
    }
    medians.toArray
  }
}

def calculateStdDevsBySpecies(data: List[Array[Double]]): Map[String, Array[Double]] = {
  val speciesData = separateDataBySpecies(data)
  speciesData.mapValues { data =>
    val stdDevs = for (i <- data.head.indices) yield {
      val col = data.map(_(i))
      val mean = col.sum / col.length
      sqrt(col.map(x => math.pow(x - mean, 2)).sum / col.length)
    }
    stdDevs.toArray
  }
}

def separateDataBySpecies(data: List[Array[Double]]): Map[String, List[Array[Double]]] = {
  val speciesData = ListBuffer.empty[(String, Array[Double])]
  for (row <- data) {
    val species = row(4).toInt match {
      case 0 => "Iris-setosa"
      case 1 => "Iris-versicolor"
      case 2 => "Iris-virginica"
    }
    speciesData += ((species, row.take(4)))
  }
  speciesData.groupBy(_._1).mapValues(_.map(_._2).toList)
}

val speciesMeans = calculateMeansBySpecies(irisData.toList)
val speciesMedians = calculateMediansBySpecies(irisData.toList)
val speciesStdDevs = calculateStdDevsBySpecies(irisData.toList)

// Combine the mean, median, and standard deviation into a single Map
val speciesStats = Map("Mean" -> speciesMeans, "Median" -> speciesMedians, "Std Dev" -> speciesStdDevs)

// Print out the mean, median, and standard deviation of each variable for each species
speciesStats.foreach { case (statType, stats) =>
  println(statType)
  stats.foreach { case (species, values) =>
    println(s"$species: ${values.mkString(", ")}")
  }
  println()
}

// Calculate the correlation matrix between each pair of variables
val corMatrix = DenseMatrix(irisData.toArray:_*).t *

File: iris.jl
using DelimitedFiles

# Load the Iris dataset into a Matrix
irisData = readdlm("/path/to/iris.csv", ',', skipstart=1)

# Calculate the mean, median, and standard deviation of each variable for each species
function calculateMeansBySpecies(data)
    speciesMeans = Dict{String, Vector{Float64}}()
    speciesData = separateDataBySpecies(data)
    for species in keys(speciesData)
        data = speciesData[species]
        means = Float64[]
        for i in 1:size(data, 2)
            col = data[:, i]
            meanValue = mean(col)
            push!(means, meanValue)
        end
        speciesMeans[species] = means
    end
    return speciesMeans
end

function calculateMediansBySpecies(data)
    speciesMedians = Dict{String, Vector{Float64}}()
    speciesData = separateDataBySpecies(data)
    for species in keys(speciesData)
        data = speciesData[species]
        medians = Float64[]
        for i in 1:size(data, 2)
            col = data[:, i]
            medianValue = median(col)
            push!(medians, medianValue)
        end
        speciesMedians[species] = medians
    end
    return speciesMedians
end

function calculateStdDevsBySpecies(data)
    speciesStdDevs = Dict{String, Vector{Float64}}()
    speciesData = separateDataBySpecies(data)
    for species in keys(speciesData)
        data = speciesData[species]
        stdDevs = Float64[]
        for i in 1:size(data, 2)
            col = data[:, i]
            stdDevValue = std(col)
            push!(stdDevs, stdDevValue)
        end
        speciesStdDevs[species] = stdDevs
    end
    return speciesStdDevs
end

function separateDataBySpecies(data)
    speciesData = Dict{String, Matrix{Float64}}()
    for row in eachrow(data)
        species = row[end]
        if !(species in keys(speciesData))
            speciesData[species] = Matrix{Float64}()
        end
        push!(speciesData[species], row[1:end-1])
    end
    return speciesData
end

speciesMeans = calculateMeansBySpecies(irisData)
speciesMedians = calculateMediansBySpecies(irisData)
speciesStdDevs = calculateStdDevsBySpecies(irisData)

# Combine the mean, median, and standard deviation into a single Dict
speciesStats = Dict("Mean" => speciesMeans, "Median" => speciesMedians, "Std Dev" => speciesStdDevs)

# Print out the mean, median, and standard deviation of each variable for each species
for (statType, stats) in speciesStats
    println(statType)
    for (species, values) in stats
        println("$species: $values")
    end
    println()
end

# Calculate the correlation matrix between each pair of variables
corMatrix = cor(irisData[:, 1:end-1])

# Print out the correlation matrix
println("Correlation Matrix")
println(corMatrix)

# TODO: Create a scatter plot matrix of the variables
